com.unity.multiplayer.mlapi/Runtime/Core/RPCQueue/RPCQueueManager.cs

```diff
@@ -0,0 +1,549 @@
+            }
+
+            //The total size of the frame is the last known position of the stream
+            queueHistoryItem.TotalSize = (uint)queueHistoryItem.QueueStream.Position;
+
+            long CurrentPosition = queueHistoryItem.QueueStream.Position;
+            ulong BitPosition = queueHistoryItem.QueueStream.BitPosition;
+
+            //////////////////////////////////////////////////////////////
+            //>>>> REPOSITIONING STREAM TO RPC MESSAGE SIZE LOCATION <<<<
+            //////////////////////////////////////////////////////////////
+            queueHistoryItem.QueueStream.Position = queueHistoryItem.GetCurrentMarkedPosition();
+
+            Int64 MSGSize = (Int64)(queueHistoryItem.TotalSize - (queueHistoryItem.GetCurrentMarkedPosition() + 8));
+            //Write the actual size of the RPC message
+            queueHistoryItem.QueueWriter.WriteInt64(MSGSize);
+
+            //////////////////////////////////////////////////////////////
+            //<<<< REPOSITIONING STREAM BACK TO THE CURRENT TAIL >>>>
+            //////////////////////////////////////////////////////////////
+            queueHistoryItem.QueueStream.Position = CurrentPosition;
+            queueHistoryItem.QueueStream.BitPosition = BitPosition;
+
+            //Add the packed size to the offsets for parsing over various entries
+            queueHistoryItem.QueueItemOffsets.Add((uint)queueHistoryItem.QueueStream.Position);
+        }
+
+        /// <summary>
+        /// GetCurrentQueueHistoryFrame
+        /// Gets the current queue history frame (inbound or outbound)
+        /// </summary>
+        /// <param name="frameType">inbound or outbound</param>
+        /// <returns>QueueHistoryFrame or null</returns>
+        private QueueHistoryFrame GetCurrentQueueHistoryFrame(QueueHistoryFrame.QueueFrameType frameType)
+        {
+            int StreamBufferIndex = GetStreamBufferIndex(frameType);
+
+            if (!QueueHistory.ContainsKey(QueueHistoryFrame.QueueFrameType.Outbound))
+            {
+                UnityEngine.Debug.LogError("You must initialize the RPCQueueManager before using MLAPI!");
+                return null;
+            }
+
+            if (!QueueHistory[QueueHistoryFrame.QueueFrameType.Outbound].ContainsKey(StreamBufferIndex))
+            {
+                UnityEngine.Debug.LogError("RPCQueueManager " + frameType + " queue stream buffer index out of range! [" + StreamBufferIndex + "]");
+                return null;
+            }
+
+            return QueueHistory[frameType][StreamBufferIndex];
+        }
+
+
+        /// <summary>
+        /// LoopbackSendFrame  - WIP
+        /// Will copy the contents of the current outbound QueueHistoryFrame to the current inbound QueueHistoryFrame
+        /// </summary>
+        public void LoopbackSendFrame()
```

Comment:
I might be being silly here but can't we just queue RPC into inbound queue rather than outbound queue right on the spot rather than doing this at all?

Reply:
You could, but this is a tool used to check that *both* the outbound and inbound queue paths are working properly.
So, you want the Rpc entries to be submitted to the outbound queue and then at the point of when they are being sent just migrate them all to the inbound queue.  This tests both sides of the egress and ingress paths.